src/betterproto/streams/__init__.py

```diff
@@ -0,0 +1,74 @@
+from .. import Message, encode_varint
+from io import BufferedReader, BufferedWriter
+from typing import Iterable, Optional
+from dataclasses import dataclass
+
+
+@dataclass(frozen=True)
+class CodedOutputStream:
+    stream: BufferedWriter
+
+    def write_varint(self, value: int) -> None:
+        """Write a single varint value to the stream."""
+        self.stream.write(encode_varint(value))
+
+    def write_message(self, message: Message) -> None:
+        """Write a serialized message to the stream."""
+        self.stream.write(bytes(message))
+
+    def write_messages(self, messages: Iterable[Message]) -> None:
+        """Write multiple messages to the stream."""
+        for message in messages:
+            message_bytes = bytes(message)
+
+            # Write the size of the message to delimit it, enabling parsing
+            self.write_varint(len(message_bytes))
+            self.stream.write(message_bytes)
```

Comment:
These two APIs would be easy to mis-use.

The `write_message` does not delimit the message but `write_messages` does.

However, I would actually suggest that the API be reworked to align with the [`io`](https://github.com/python/typeshed/blob/797779d222b492d98065dafb30d828d3013bca8b/stdlib/io.pyi#L121C58-L121C58) API does: just a `write` function:

```py
@dataclass(frozen=True)
class CodedOutputStream:
    stream: BufferedWriter

    def __write(self, message: Message) -> int:
        buffer = bytes(message)
        varint = encode_varint(len(buffer))
        return self.stream.write(varint) + self.stream.write(buffer)

    def write(self, __buffer: Union[Message, Iterable[Message]]) -> int:
        if isinstance(__buffer, Message):
            return self.__write(__buffer)

        return sum(self.__write(m) for m in __buffer)
```

You may want to add an assertion that the `stream` is correct when initialised:

```py
    def __post_init__(self) -> None:
        if "b" not in self.stream.mode:
            raise ValueError(...)
        if not self.stream.readable:
            raise ValueError(...)
```

Depending on how you see this class being used, you may want to enable it as a Context manager with `__enter__`/`__exit__` so one can do:

```py
with CodedOutputStream(path.open("wb")) as stream:
    stream.write(messages)
```

You could do a similar thing with the below `CodedInputStream`.

Reply:
This has been moved into the main module, and the new functions are `Message.write_to` and `Message.write_delimited_to`. The former just does this and writes the message alone to the stream, but the latter writes the message to the stream prefixed with the varint size. This means you have both options, and means it is in line with what at least the Java Protobuf API does with Message writing, but it does also mean that reading and writing messages doesn't directly handle iteration. Iteratively reading and writing should still be trivial using the new methods, but I wasn't sure if it would be a good idea or not to add separate functions just for handling multiple messages in/out on top of this - let me know what you think.

I'll also add checks like suggested for whether the stream is readable/writable/binary to the new functions, and the new versions also have typehints to accept `BinaryIO` which should also make it a bit clearer to users what they should pass in.